lib/Parse/ParseStmt.cpp

```diff
@@ -1228,19 +1235,34 @@ validateAvailabilitySpecList(Parser &P,
       P.diagnose(InsertWildcardLoc, diag::attr_availability_wildcard_in_macro);
     }
   } else if (!HasOtherPlatformSpec) {
-    SourceLoc InsertWildcardLoc = P.PreviousLoc;
-    P.diagnose(InsertWildcardLoc, diag::availability_query_wildcard_required)
+    if (isUnavailability) {
+      // Imply * in #unavailable statements
+      auto wldcrd = new (P.Context) OtherPlatformAvailabilitySpec(SourceLoc());
+      RecognizedSpecs.push_back(wldcrd);
+    } else {
+      SourceLoc InsertWildcardLoc = P.PreviousLoc;
+      P.diagnose(InsertWildcardLoc, diag::availability_query_wildcard_required)
         .fixItInsertAfter(InsertWildcardLoc, ", *");
+    }
```

Comment:
There is a minor bug-fix opportunity here as well as opportunity to make the code clearer.

The bug is that with an availability macro, this code is assuming that `*` is the last entry (which is very likely but not strictly guaranteed). So if you make a command like:

```
xcrun swiftc -Xfrontend -define-availability -Xfrontend "SwiftStdlib 5.5:macOS 9999, *,
watchOS 9999, tvOS 9999" tmp.swift
-define-availability argument:1:51: error: future platforms identified by '*' cannot be used in an availability macro definition
SwiftStdlib 5.5:macOS 9999, *, watchOS 9999, tvOS 9999
                                                  ^
```
the caret is pointing to the wrong token. Instead that check should be happening inside the loop similar to where the unavailable + `*` check is happening right now.

The Unavailabile and Macro cases are similar in that we need to diagnose an error if a wildcard is present. The difference is that in the Unavailable case, you need to provide a fix-it as well.

Here, the nested-if makes things a bit trickier to follow IMO. It would be nicer to have a switch:

```
if (!HasOtherPlatformSpec) { 
  switch (specSource) {
  case AvailabilitySpecSource::Macro: break;
  case AvailabilitySpecSource::Available: /* issue error and fix-it */
  case AvailabilitySpecSource::Unavailable: /* store new spec */
  }
}
```